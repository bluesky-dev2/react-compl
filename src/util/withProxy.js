import React, { PureComponent } from 'react';
import { render, unmountComponentAtNode } from 'react-dom';

export default function(Component) {
  return class WithProxy extends PureComponent {
    node = null;

    componentDidMount() {
      this.appendToBody(this.props);
    }

    componentWillUnmount() {
      this.removeFromBody();
    }

    componentWillReceiveProps(nextProps) {
      this.removeFromBody();
      this.appendToBody(nextProps);
    }

    removeFromBody() {
      unmountComponentAtNode(this.node);
      document.body.removeChild(this.node);
    }

    appendToBody(props) {
      const container = document.createElement('div');
      this.node = document.body.appendChild(container);
      render(<Component {...props} key={props.id} />, this.node);
    }

    render() {
      return null;
    }
  };
}
