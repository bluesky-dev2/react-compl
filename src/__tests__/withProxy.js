/* eslint-env jest */
import React from 'react';
import { shallow, mount } from 'enzyme';
import toJson from 'enzyme-to-json';

import withProxy from './../util/withProxy';

beforeEach(() => (document.body.innerHTML = ''));

describe('withProxy', () => {
  it('Wrapped component should render null', () => {
    const Foo = () => <div>Foobar</div>;
    const Component = withProxy(Foo);
    const renderedComponent = shallow(<Component />);
    expect(renderedComponent.html()).toBe(null);
  });

  it('Should remove the Html node when unmount', () => {
    const Foo = () => <div>Foobar</div>;
    const Component = withProxy(Foo);
    const renderedComponent = mount(<Component />);
    expect(document.body.innerHTML).toBe('<div><div>Foobar</div></div>');
    renderedComponent.unmount();
    expect(document.body.innerHTML.length).toBe(0);
  });

  it('Should update the component when receive new props', () => {
    const Foo = ({ param }) => <div>Foobar {param}</div>;
    const Component = withProxy(Foo);
    const renderedComponent = mount(<Component param="coucou" />);
    renderedComponent.setProps({ param: 'plop' });

    expect(toJson(renderedComponent)).toMatchSnapshot();
  });
});
