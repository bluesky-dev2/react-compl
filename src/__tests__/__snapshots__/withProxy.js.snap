// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withProxy Should update the component when receive new props 1`] = `
<WithProxy
  param="plop"
/>
`;
